Fixed port connection breaking when switching to Play Mode and added some QOL improvements to make BlueGraph more extensible

thechayed opened the pull request:
Likely due to recent updates to Unity, graphs would break when switching to Play Mode. To fix this issue, I added a Port Connection Cache and unique ID for Node Ports, allowing them to be reconnected when needed. This can be done externally, but is also done iteratively in the GraphEditorWindow while it is open. 

Some QOL improvements were added to the Graph as well that makes it easier to extend for custom tooling.

thechayed committed:
Small changes to make the system more modular

thechayed committed:
Added a port connection Cache for the Graph that is used to reconstruct connections when needed, graphs no longer break when switching between Edit/Play Modes

thechayed committed:
Add GetNodeArray() to Graph to get nodes previously cached by their Type

thechayed committed:
Changed name of GetNodesArray() to GetCachedNodes() for transparency

AncientPixel commented:
BlueGraph is a excellent package I've integrated with my project but I don't think it's actively being maintained at the moment, might have to fork it. But thanks for your efforts, I hope someone approves your PR. :)

thechayed commented:
> BlueGraph is a excellent package I've integrated with my project but I don't think it's actively being maintained at the moment, might have to fork it. But thanks for your efforts, I hope someone approves your PR. :)

It really is~! I personally like it more than Unity’s GTK. it’s way easier to extend, even if it’s not the direction Unity’s going.
I’m super happy I could help a little with the changes I made, even if the PR doesn’t get approved ♥
If you want, I’ll keep tinkering with my fork and add anything fun or useful I come across~

McManning commented:
> BlueGraph is a excellent package I've integrated with my project but I don't think it's actively being maintained at the moment, might have to fork it. But thanks for your efforts, I hope someone approves your PR. :)

Sorry about that, I don't have the capacity at the moment to actively maintain this project outside of PR review.

I really do appreciate this fix, I'll get some time set aside to review this today or tomorrow.

McManning commented on Editor/Resources/BlueGraphEditor/NodeView Hovered.uss:
Might want to drop these empty `NodeView Hovered.uss/meta` files from your PR

thechayed replied:
Will do!

McManning commented on Runtime/Node.cs:
```diff
@@ -133,6 +150,9 @@ public void Validate()
             OnValidateEvent?.Invoke();
+
+            if (GraphID == "!")
+                return;
```
These couple lines with the magic `"!"` ID don't seem necessary here.

thechayed replied:
🙈 thanks for catching it! I’ll fix it up.

McManning commented on Runtime/Utils/SerializedDictionary.cs:
```diff
@@ -0,0 +1,287 @@
+using JetBrains.Annotations;
```
Will this JetBrains import lead to issues for non-JetBrains users?

thechayed replied:
No, I don't actually use Jetbrains myself, but Visual Studio keeps automatically adding it as a line in my code when I copy/paste things, which I did with this SerializedDictionary which is what I've used for my other project. I should look into how to disable that functionality.

McManning commented on Runtime/Utils/SerializedDictionary.cs:
```diff
@@ -0,0 +1,287 @@
+    namespace Remedy.Framework
+    {
+        [Serializable]
+        public class SerializableDictionary<TKey, TValue> : IDictionary<TKey, TValue>, ISerializationCallbackReceiver
```
I'd say we want to drop the nested `Remedy.Framework` namespace to clean this up a bit. If this is from their code and we want to still maintain credit, I'd probably just have some authorship info in the SerializableDictionary's docblock. 

Small nit as well but I'd rather be consistent with class name and .cs filename. Both as `SerializableDictionary` would be good.

thechayed replied:
Sorry about that, I forgot to change the namespace when migrating the code over to BlueGraph, fixing it now..

McManning commented on Runtime/Attributes.cs:
```diff
@@ -229,4 +232,20 @@ public CustomNodeViewAttribute(Type nodeType)
+    /// Tells the Node Caching system to cache the Node to a specific type. 
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
+    public class CacheToAttribute : Attribute
```
Do you have an example of how would you use this? So we can add something to the Wiki docs.

thechayed replied:
Okay, sorry for all the edits! I realized my original response was probably too complex.

I added a Node Cache that caches nodes by their type, or by a desired base type/interface using CacheToAttribute. Lookups using the cache are O(1) with no extra overhead compared to the previous methods, and it returns an array for traversal, which is more efficient at runtime.

CacheTo lets you group nodes together by a base type or interface, so you can declare functionality based on a shared type rather than the concrete node type.

thechayed committed:
Fix minor issues from PR feedback

- Remove unnecessary magic '!' check in Node.Validate
- Remove JetBrains namespace import and fix namespace in SerializableDictionary
- Ensure class name matches filename for SerializableDictionary
- Clarify usage of CacheToAttribute and node caching logic

McManning commented:
Started running this in my samples project (https://github.com/McManning/BlueGraph-Samples). Got it to work, sans a few minor adjustments I had to make:

1. The shorthand syntax `= new();` and `??= new();` aren't supported by 2019.3 since it ran an older version C#. I'm not against bumping minimum supported version up - given that was more than 5 years ago at this point - but since there's not a *lot* that has changed in this PR that demands newer Unity support we could just do something like the following:

```cs

        [SerializeField]
        private SerializableDictionary<SerializableType, Node[]> nodeByTypeCache;
        /// <summary>
        /// The nodes Cached in the Graph, by their type so they can be easily queried.
        /// </summary>
        protected SerializableDictionary<SerializableType, Node[]> NodesByTypeCache
        {
            get
            {
                if (nodeByTypeCache == null)
                {
                    nodeByTypeCache = new SerializableDictionary<SerializableType, Node[]>();
                }

                return nodeByTypeCache;
            }
        }

        /// <summary>
        /// The cached connections between nodes from the Schematics Editor in a Dictionary that allows them to be reconstructed if they're lost.
        /// </summary>
        [SerializeField]
        [HideInInspector]
        internal SerializableDictionary<string, string[]> portConnectionCache = new SerializableDictionary<string, string[]>();

        /// <summary>
        /// A Cache pairing Port IDs to their Ports that is concstructed at Runtime to improve Port Lookup time
        /// </summary>
        [SerializeField]
        [HideInInspector]
        private SerializableDictionary<string, Port> runtimePortCache = new SerializableDictionary<string, Port>();
```

---

2. There was a meta file I needed to delete from the branch, otherwise Unity would complain in a loop 

<img width="1094" height="65" alt="image" src="https://github.com/user-attachments/assets/aa57ce0c-55d5-403f-acb6-0e589dfc7a40" />

---

3. Couple minor USS bugs with node headers losing color and the inline editors missing a background:
<img width="545" height="874" alt="image" src="https://github.com/user-attachments/assets/8b16e5fc-3d51-4f27-ace0-22f01019ae23" />

Port colors also seem to be broken versus what it should be (https://github.com/McManning/BlueGraph-Samples?tab=readme-ov-file#execution-flow):

<img width="683" height="319" alt="image" src="https://github.com/user-attachments/assets/1289ece8-780b-448d-9e2d-c2a2ab68758b" />

Still digging into what's causing this one. May not have necessarily been this PR. The USS issues aren't a blocker here for me though, I can push a patch to fix USS easily enough after merging this. 

I think just the extra `.meta` file causing warnings and whether or not we want to patch the syntax a bit to maintain older Unity support that would be left to do here. Looks good otherwise, and super appreciated!

McManning commented:
> Still digging into what's causing this one. May not have necessarily been this PR. The USS issues aren't a blocker here for me though, I can push a patch to fix USS easily enough after merging this.

This looks like it's probably Unity compatibility issues as well while I test under 2019:

```
StyleSheet import: type=Validation, code=InvalidProperty file=Packages/com.github.mcmanning.bluegraph/Editor/Resources/BlueGraphEditor/NodeView.uss:148
    Expected (flex-start | flex-end | center | stretch | auto) but found 'left'
    align-self: left
    
StyleSheet import: type=Validation, code=InvalidProperty file=Packages/com.github.mcmanning.bluegraph/Editor/Resources/BlueGraphEditor/NodeView.uss:156
    Unknown property 'transition' (did you mean 'position'?)
    transition: height 0.5s
```

If these USS properties were introduced soonish after 2019, I'd probably be fine bumping the minimum supported version of this addon.
